src/dune_util/located_stringlike.mli

```diff
@@ -0,0 +1,6 @@
+open! Stdune
+
+(** Implement [Stringlike_intf.S] with a value accompanied by a [Loc.t] for
+    string that carry around their own source location for easier error
+    reporting. *)
+module Make (S : Stringlike_intf.S_base) : Stringlike_intf.S with type t = Loc.t * S.t
```

Comment:
We've had something like this in the past, but it turned out to be problematic. The issue is that it makes it too easy not to think about locations whens doing comparisons. When we do `Foo.equal` we should really think about whether we need to take the location into account when comparing identifiers. For that reason we don't include the locations in such type definitions.

Reply:
What if we changed the functor to take an additional argument specifying whether comparisons should include the loc or not so you're forced to consciously make that decision when instantiating it?

Alternatively we could just include the loc in `Package_name.t` and treat it as a special (and hopefully temporary) case.

Reply:
Isn't it easy enough to just pass `Loc.t * Package_name.t` wherever you need the location? That's what we do everywhere else (library names, repository names, etc.)

Reply:
The `Loc.t` is needed whenever we convert from a `Package_name.t` to a `Package_name.Opam_compatible.t` and I figured it would be a big refactor to add `Loc.t` to all the relevant places.

Reply:
Tbh, it doesn't seem so bad - I've done it just about everywhere else without paying attention to it. In any case, I'm fine with this code as well. Although CI says it's not working yet.

Reply:
Oh btw, could you move this module closer to where it's being used. `Dune_util` is being used by the engine so ideally we wouldn't anything that isn't needed there to avoid syncing issues. I realize it already has some stuff that probably shouldn't live there, but we should avoid adding more.

Reply:
I'll do an experiment to see how bad the refactoring would actually be and if it turns out fine then I'll just delete this module. Otherwise yep I'll move it to dune_lang.

Reply:
It would be a big refactor to add `Loc.t` to all the places where it's needed so I've instead moved `Located_stringlike` into the `dune_lang` library.